source/basic_skills.rst

```diff
@@ -886,18 +887,410 @@ The proofs we have carried out in this section provide some hints.
+    end
+
+To try out ``library_search`` in this example,
+delete the ``exact`` command and uncomment the previous line.
+Using these tricks,
+see if you can find what you need to do the
+next example:
+
+.. code-block:: lean
+
+    import import analysis.special_functions.exp_log
+    import tactic
+
+    open real
+
+    variables a b c : ℝ
+
+    -- BEGIN
+    example (h : a ≤ b) : c - exp b ≤ c - exp a :=
+    begin
+      sorry
+    end
+    -- END
+
+Here is another example of an inequality:
+
+.. code-block:: lean
+
+    import data.real.basic tactic
+
+    variables a b : ℝ
+
+    -- BEGIN
+    example : 2*a*b ≤ a^2 + b^2 :=
+    begin
+      have h : 0 ≤ a^2 - 2*a*b + b^2,
+      calc
+        a^2 - 2*a*b + b^2 = (a - b)^2     : by ring
+        ... ≥ 0                           : by apply pow_two_nonneg,
+      calc
+        2*a*b
+            = 2*a*b + 0                   : by ring
+        ... ≤ 2*a*b + (a^2 - 2*a*b + b^2) : add_le_add (le_refl _) h
+        ... = a^2 + b^2                   : by ring
+    end
+    -- END
+
+Mathlib tends to put spaces around binary operations like ``*`` and ``^``,
+but in this example, the more compressed format increases readability.
+There are a number of things worth noticing in this example.
+First, an expression ``s ≥ t`` is definitionally equivalent to ``t ≤ s``.
+In principle, this means one should be able to use them interchangeably.
+But some of Lean's automation does not recognize the equivalence,
+so Mathlib tends to favor ``≤`` over ``≥``.
+Second, we have used the ``ring`` tactic extensively.
+It is a real timesaver!
+Finally, notice that in the second line of the
+second ``calc`` proof,
+instead of writing ``by exact add_le_add (le_refl _) h``,
+we can simply write the proof term ``add_le_add (le_refl _) h``.
+
+In fact, the only cleverness in the proof above is figuring
+out the hypothesis ``h``.
+Once we have it, the second calculation involves only
+linear arithmetic, and ``linarith`` can handle it:
+
+.. code-block:: lean
+
+    import data.real.basic tactic
+
+    variables a b : ℝ
+
+    -- BEGIN
+    example : 2*a*b ≤ a^2 + b^2 :=
+    begin
+      have h : 0 ≤ a^2 - 2*a*b + b^2,
+      calc
+        a^2 - 2*a*b + b^2 = (a - b)^2 : by ring
+        ... ≥ 0                       : by apply pow_two_nonneg,
+      linarith
+    end
+    -- END
+
+How nice! We challenge you to use these ideas to prove the
+following theorem. You can use the theorem ``abs_le_of_le_of_neg_le``.
+
+.. code-block:: lean
+
+    import data.real.basic tactic
+
+    variables a b : ℝ
+
+    -- BEGIN
+    example : abs (a*b) ≤ (a^2 + b^2) / 2 :=
+    sorry
+
+    #check abs_le_of_le_of_neg_le
+    -- END
+
+If you managed to solve this, congratulations!
+You are well on your way to becoming a master formalizer.
+
+
+.. _using_more_theorems_and_lemmas:
+
+Using More Theorems and Lemmas
+------------------------------
+
+[I got tired of writing, so I decided to give the reader a break
```

Comment:
Is this the place to discuss optional arguments to `linarith`? In particular giving a list of expressions to `linarith` often allows to avoid stating intermediate fact. This makes proofs less readable but more compact.

Reply:
Good idea. One of the examples I give the reader involves subtraction and exp. I tell them to do it by hand, to practice finding theorems. But I will also tell them to do it with linarith (after showing a similar example).